factor_report.php

```diff
@@ -0,0 +1,95 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+/**
+ * Reporting page for each factor vs auth type
+ *
+ * @package   tool_mfa
+ * @author    Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright 2019 Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require_once(__DIR__ . '/../../../config.php');
+require_once($CFG->libdir.'/adminlib.php');
+
+admin_externalpage_setup('factorreport');
+
+$PAGE->set_title(get_string('factorreport', 'tool_mfa'));
+$PAGE->set_heading(get_string('factorreport', 'tool_mfa'));
+
+$factors = \tool_mfa\plugininfo\factor::get_enabled_factors();
+
+// Setup 2 arrays, one with internal names, one pretty.
+$columns = array(get_string('tableheader', 'tool_mfa'));
+$displaynames = $columns;
+$colclasses = array('center');
+foreach ($factors as $factor) {
+    $columns[] = $factor->name;
+    $displaynames[] = get_string('pluginname', 'factor_'.$factor->name);
+    $colclasses[] = 'right';
+}
+
+$table = new \html_table();
+$table->head = $displaynames;
+$table->align = $colclasses;
+
+// Total row.
+$totalrow = array(get_string('total'));
+foreach ($columns as $column) {
+    if ($column == get_string('tableheader', 'tool_mfa')) {
+        continue;
+    }
+    $sql = 'SELECT COUNT(DISTINCT tm.userid)
```

Comment:
this is probably fine as is, but I'm curious if this could be rolled up into a single sql for the lot. I suspect this may be a little hog on a large site.

Reply:
UX wise I'd move the totals to the bottom